### ISSUE TITLE
System.ArgumentException: The output char buffer is too small to contain the decoded characters, encoding 'Unicode (UTF-8)'

### ISSUE THREAD

**Issue by mini890** [villermen/runescape-cache-tools#47]
I don't know if I'm doing anything wrong but when I try to execute the soundtracks from the cache it throws me an error, I'm executing

`rsct.exe --output-directory out --soundtrack-combine`

And it throws me the following error:

`System.ArgumentException: The output char buffer is too small to contain the decoded characters, encoding 'Unicode (UTF-8)' fallback 'System.Text.DecoderReplacementFallback'.
Parameter name: chars

in System.Text.Encoding.ThrowCharsOverflow()

in System.Text.Encoding.ThrowCharsOverflow(DecoderNLS decoder, Boolean nothingDecoded)

in System.Text.UTF8Encoding.GetChars(Byte* bytes, Int32 byteCount, Char* chars, Int32 charCount, DecoderNLS baseDecoder)

in System.Text.DecoderNLS.GetChars(Byte* bytes, Int32 byteCount, Char* chars, Int32 charCount, Boolean flush)

in System.Text.DecoderNLS.GetChars(Byte[] bytes, Int32 byteIndex, Int32 byteCount, Char[] chars, Int32 charIndex, Boolean flush)

in System.Text.DecoderNLS.GetChars(Byte[] bytes, Int32 byteIndex, Int32 byteCount, Char[] chars, Int32 charIndex)

in System.IO.BinaryReader.InternalReadOneChar()

in System.IO.BinaryReader.Read()

in System.IO.BinaryReader.ReadChar()

in Villermen.RuneScapeCacheTools.Extensions.BinaryStreamExtensions.ReadNullTerminatedString(BinaryReader reader)

in Villermen.RuneScapeCacheTools.Enums.EnumFile.Decode(Byte[] data)

in Villermen.RuneScapeCacheTools.Cache.CacheFile.FromDataFile(DataCacheFile dataFile)

in Villermen.RuneScapeCacheTools.Cache.CacheBase.GetFile[T](Index index, Int32 fileId, Int32 entryId)

in Villermen.RuneScapeCacheTools.Audio.Soundtrack.GetTrackNames(Boolean includeUnnamed)

in Villermen.RuneScapeCacheTools.Audio.Soundtrack.Extract(Boolean overwriteExisting, Boolean lossless, Boolean includeUnnamed, String[] nameFilters)

in Villermen.RuneScapeCacheTools.CLI.Program.CombineSoundtrack()

in Villermen.RuneScapeCacheTools.CLI.Program.Main(String[] args)
`
Sorry if it's not exactly the same names, but my OS it's in Portuguese and it throws me the error in Portuguese.

**Comment by villermen** [comment]
Thanks for the detailed report. It's most definitely not your fault but
mine. I'll investigate when I wake up. (I'm fast asleep at the moment.)

**Comment by mini890** [comment]
It's fine. Anything I can do to help (if it's within my skills), just say.

**Comment by villermen** [comment]
At closer inspection I might have overlooked something that changes based on OS in reading the list of tracks. It might however be unrelated and have something to do with your specific cache. Does it happen every time when you run that specific command? Is it possible for you to send me your cache directory (all .idx# and the .dat2 file usually located in C:\Users\[yourusername]\jagexcache\runescape\LIVE)? That way I can verify it's not the cache (which I hope it is not). Either way I have to fix something in the code in order for it to work correctly.

**Comment by mini890** [comment]
I can. Where should I send you the files?

**Comment by villermen** [comment]
It's LIVE. The other one might be a modded/private server cache or something. You can zip them and drag/drop them in a comment, or mail me them, puush them. Whichever way works best for you.

**Comment by mini890** [comment]
Ok. I'm compressing. I take it you shouldn't have any problem with 7z?

**Comment by villermen** [comment]
I can decompress anything =)

**Comment by mini890** [comment]
I emailed you the cache. Sorry for the delay. Slow Internet Connection

**Comment by villermen** [comment]
Hehe, it's really weird to suddenly see Portuguese names pop up where I am used to seeing English names.

![Portuguese track names](https://viller.men/owncloud/s/FSlIdPtjaqQuVZg/download)

I did encounter the same issue you had the first time I ran though. It is due to the wrong character set being used. That only you encountered it is purely due to special characters that are used in the Portuguese track names. I have written a fix (e9a06e6e60d42a04b0468352a5c02fe76e66a7d9) and will release it as soon as it passes the tests.

Technical mumbo jumbo (mostly a sidenote for me): I was using UTF-8 for reading and writing strings, but actually ISO-8859-1 should have been used. UTF-8 uses the last bit in a byte to signal that another byte is coming, whereas ISO-8859-1 fills them in with extra characters (ISO-8859-1 is a single-byte encoding whereas UTF-8 supports a variable byte-length). When reading a certain character, the reader expected another byte to come, but there were no bytes left.
